compiler.py

```diff
@@ -0,0 +1,637 @@
+
+        elif method == 'flush':
+            if len(node.args) == 0:
+                self._ins.append(f'drawflush display1')
+            elif len(node.args) == 1:
+                value = self.as_value(node.args[0])
+                self._ins.append(f'drawflush {value}')
+            else:
+                raise CompilerError(ERR_BAD_SYSCALL_ARGS, node)
+
+        else:
+            raise CompilerError(ERR_UNSUPPORTED_SYSCALL, node)
+
+    def emit_control_syscall(self, node: ast.Call):
+        """This will be called by visit_Expr when it encounters a Control(...) command."""
+
+        # TODO may want to reimplement much of this in a more object-oriented fashion, e.g., ripple1.shoot(...)
+        #      rather than Control.shoot( ripple1, ... )
+
+        method = node.func.attr
+        if method == 'enabled':
+            if len(node.args) != 2:
+                raise CompilerError(ERR_BAD_SYSCALL_ARGS, node)
+
+            link, enabled = map(self.as_value, node.args)
+            self._ins.append(f'control enabled {link} {enabled}')
+        elif method == 'shoot':
+            if len(node.args) == 3:
+                link, x, y, enabled = *map(self.as_value, node.args), 1
+            elif len(node.args) == 4:
+                link, x, y, enabled = map(self.as_value, node.args)
+            else:
+                raise CompilerError(ERR_BAD_SYSCALL_ARGS, node)
+
+            self._ins.append(f'control shoot {link} {x} {y} {enabled}')
+        elif method == 'ceasefire':
+            if len(node.args) != 1:
+                raise CompilerError(ERR_BAD_SYSCALL_ARGS, node)
+
+            link = self.as_value(node.args[0])
+            self._ins.append(f'control shoot {link} 0 0 0')
+        else:
+            raise CompilerError(ERR_UNSUPPORTED_SYSCALL, node)
+
+    def as_value(self, node):
+        """This will be called by various visit methods when they encounter something to treat as an atomic expression,
+           e.g. for the y and 3 in x=y+3.  This coerces the given node to a string that can appear in mlog code."""
+
+        # TODO figure out balance of labor between this and other things in evaluating complex expressions
+
+        if isinstance(node, ast.Constant):  # constants include True, 1.5, and "yarn"
+            if isinstance(node.value, bool):
+                return ('false', 'true')[node.value]
+            elif isinstance(node.value, (int, float)):
+                return str(node.value)
+            elif isinstance(node.value, str):
+                return '"' + ''.join(c for c in node.value if c >= ' ' and c != '"') + '"' #scrub quotes and control characters from string
+            else:
+                raise CompilerError(ERR_COMPLEX_VALUE, node)
+
+        elif isinstance(node, ast.Name):
+            assert isinstance(node.ctx, ast.Load)
+            # TODO should this raise a CompilerError like most other errors do?
+            return node.id
+            # TODO figure out what this does???

```

Comment:
When we encounter `name`, we just return that `name`. mlog works with variable names anywhere, so references to Python variables don't really need to be translated.